### ISSUE TITLE
make it easier to fine-tune when the pretrained model is only part of the final model

### ISSUE THREAD

**Issue by joelgrus** [allenai/allennlp#1607]
both @nelson-liu and @csbhagav have needed to do this recently.

in one case, we needed to train a model that used a pretrained `Seq2SeqEncoder`. in another we needed to train just a classification layer at the end of a pretrained `Model`.

we came up with the following hacky workaround:

(1) have the new model take as constructor parameters `pretrained: Model` (or `pretrained: Seq2SeqEncoder`) and also `pretrained_file: str`
(2) respecify the pretrained config as a subconfig for the new model so that it gets instantiated from params correctly
(3) in the new model constructor, if `pretrained_file` is not None, load the weights for that component from the pretrained file
(4) `load_archive` has logic that throws out config keys that are `pretrained_file` (which was put there to deal with the analogous issue for word embeddings), which means that after fine-tuning the resulting model will have all of its weights and won't need to keep around the pretrained file

we should figure out a cleaner, more general way to handle this situation

**Comment by HarshTrivedi** [comment]
@joelgrus  This would be very helpful! I had been doing something very much similar expect that in my code most logic was handled in `from_params` (*started before it was 'mostly' removed.*). And I used `train` command to do everything.

A typical scenario that I often have to handle is: I have pretrained **model A** and pretrained **model B** and now I want to create a new **model X**, but in that want to borrow some pretrained modules from A and some modules from **model B**. Also I need configurability to which modules to freeze and which modules to keep tunable when borrowed.  For **model X** ( above scenario ) the config for my code would be something like:

```
    "model": {
        "type": "model_x_type",
        "borrow_model_A": {
                "dir": "model_A_path",
                "borrow_modules": {
                    "text_field_embedder" : "transfer",
                    "encoderA1"  : "tune",
                    "encoderA2" : "tune",
                }
        },
        "borrow_model_B": {
                "dir": "model_B_path",
                "borrow_modules": {
                    "encoderX"  : "transfer",
                    "encoderY"  : "use",
                }
        },
```
I have three modes of borrowing modules: 
1. `transfer` means - only transfer and freeze those modules while model X training.
2. `tune` means transfer module parameters and tune it while model X training.
3. `use` means use module definition from pretrained file, but keep initialization based on model X's code.

Based on this config, modules were borrowed with `load_archive` in `from_params` and passed to `__init__`. From the config, regexes of borrowed modules to freeze and modules to not initialize (`prevent` regex ) were obtained and passed to `__init__` where they are handled. Rough outline:
```python
class MyModelX(Model):

    def __init__(self, ...
                 text_field_embedder: TextFieldEmbedder,
                 encoderA: Seq2VecEncoder,
                 ... 
                 encoderA: Seq2SeqEncoder,
                 nograd_regexes,
                 initializer):
               ...
               ...
               initializer(self)
               for name, parameter in self.named_parameters():
                     if any(re.search(regex, name) for regex in nograd_regexes):
                          parameter.requires_grad_(False)

    def from_params( ... ):
        ...
        transfer_model_configs = params.pop("borrow_model_A")
        modules_dict, nograd_module_names, noinit_module_names = util.borrow_modules(transfer_model_configs)
        # modules_dict is {<module_name> : borrowed_module } e.g {"encoderX": Seq2Vec }...
        ...
        noinit_regexes = [".*{}.*".format(name) for name in noinit_module_names]
        nograd_regexes = [".*{}.*".format(name) for name in nograd_module_names]
        initializer_params += [[regex, "prevent"] for regex in noinit_regexes]
        nograd_regexes = [".*{}.*".format(name) for name in nograd_module_names]
        return cls(..., **modules_dict, nograd_regexes=nograd_regexes, initializer=initializer)
```


I am sharing this scenario so that such a use case can be kept in mind while coding this feature.

PS: I am busy with my deadlines for next 5 weeks. If no body else from team has taken up by then, I would be happy to *discuss* and send pr for *refactored* version of doing this (at least skipping from_params).

**Comment by matt-gardner** [comment]
@joelgrus, it seems to me like the cleanest thing to do here is to have a special `from_archive` parameter that's allowed in the `Model.from_params` method.  This would just load the model from the given path and return it (you could imagine throwing in some `trainable`-type parameters also, to accomplish something like what @HarshTrivedi mentioned, though that's getting a bit complicated).

For training something, this seems pretty straightforward.  The place it gets complicated is when loading a model that's been saved after training in this way.  You have two choices there: just copy the archive file into the new model's archive, so you can still load everything, but then you're duplicating the model's weights, which could be wasting a whole lot of space.  The alternative is to be super fancy and add a mechanism for modifying pieces of the config between loading the original config and saving it to the archive, so that you replace the `from_params` block with the `model: {}` piece from the config inside the archive you loaded.  Seems like the first thing to do is just copy the model archive over using the existing `files_to_archive` mechanism, and make it fancier if we ever need it.

**Comment by HarshTrivedi** [comment]
How about an argument in `__init__` something like  `transfer_model: TransferModel` which is thin wrapper around `from_archive` consuming params config of roughly following type: .. ?
```
"borrow_model_B": {
    "dir": "model_B_path",
        "borrow_modules": {
            "encoderA"  : "transfer",
            "encoderB"  : "tune",
            } 
    }
```

**Comment by handsomezebra** [comment]
Another choice which I found clean is to use initializer, and use regex syntax for key mappings, something like:

```
  "initializer": [
    ["module_A.*", {"type": "pretrained", "archive_file_name": "pretrained_A_path", "key_name": "existing_module_A\\0"}],
    ["module_B.*weight.*", {"type": "pretrained", "archive_file_name": "pretrained_B_path", "module_name": "existing_module_B\\0weight\\1"}],
    ...
  ]
```

**Comment by joelgrus** [comment]
I like the initializer approach, it would likely require some tweaks to how initializers work, but it feels right

**Comment by matt-gardner** [comment]
Using the initializers feels like a much bigger re-write of how those work.  If you want to initialize whole structures instead of individual weights, using our existing mechanism is a bit of a train wreck.

We basically have a system in place that says that every module is responsible for its own initialization, but you can globally tweak things after the fact if you really want to with the "initializer" key.  The ELMo code and the embedding code are the only places where we currently have some pre-training, and they both handle the initialization from pre-trained files locally.  I'm favor of keeping this current system, because if you want to change how the initialization mechanism works to make it more amenable to initializing whole structures instead of just giving small tweaks, you're probably going to end up duplicating the structure of the other parameters inside the "initializer" key, which seems like a waste.

**Comment by matt-gardner** [comment]
Fixed by #2043.  If there are further issues, or that PR doesn't cover a desired use case, feel free to open another issue describing what's missing or wrong.
